File: smtk/extension/matplotlib/testing/python/renderMesh.py
#=============================================================================
#
#  Copyright (c) Kitware, Inc.
#  All rights reserved.
#  See LICENSE.txt for details.
#
#  This software is distributed WITHOUT ANY WARRANTY; without even
#  the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR
#  PURPOSE.  See the above copyright notice for more information.
#
#=============================================================================
import os
import sys
import unittest
import smtk
import smtk.io
import smtk.mesh
import smtk.model
import smtk.extension.matplotlib.render_mesh
import smtk.testing
from smtk.simple import *


class RenderMesh(smtk.testing.TestCase):

    def setUp(self):

        # Construct an operation manager
        self.operationManager = smtk.operation.Manager.create()

        # Register matplotlib operations to the operation manager
        smtk.extension.matplotlib.registerOperations(self.operationManager)

        # Set up the path to the test's input file
        meshFile = os.path.join(
            smtk.testing.DATA_DIR, 'mesh', '2d', 'warpedMesh.h5m')

        # Read in the mesh
        self.collection = smtk.mesh.Collection.create()
        smtk.io.readMesh(meshFile, self.collection)

    def testRenderMesh(self):

        # Create a "render mesh" operator
        renderMesh = self.operationManager.createOperator(
            'smtk.extension.matplotlib.render_mesh.RenderMesh')

        if not renderMesh:
            raise ImportError('Could not find operator \'render mesh\'')

        # Set the operator's parameters
        print renderMesh.parameters()
        renderMesh.parameters().find('mesh').setValue(self.collection.meshes())

        self.outFile = os.path.join(
            smtk.testing.TEMP_DIR, str(smtk.common.UUID.random()) + '.png')

        renderMesh.parameters().find('filename').setValue(self.outFile)

        # Execute the operator
        result = renderMesh.operate()

        if self.interactive() and self.haveVTK() and self.haveVTKExtension():
            import vtk
            self.startRenderTest()
            reader = vtk.vtkPNGReader()
            reader.SetFileName(self.outFile)
            reader.Update()
            self.addImageToScene(reader)
            cam = self.renderer.GetActiveCamera()
            self.renderer.ResetCamera()
            self.renderWindow.Render()
            self.interact()


if __name__ == '__main__':
    smtk.testing.process_arguments()
    unittest.main()

File: smtk/extension/matplotlib/testing/cxx/RenderMesh.cxx
//=========================================================================
//  Copyright (c) Kitware, Inc.
//  All rights reserved.
//  See LICENSE.txt for details.
//
//  This software is distributed WITHOUT ANY WARRANTY; without even
//  the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR
//  PURPOSE.  See the above copyright notice for more information.
//=========================================================================

#include "smtk/attribute/Attribute.h"
#include "smtk/attribute/FileItem.h"
#include "smtk/attribute/IntItem.h"

#include "smtk/common/PythonInterpreter.h"

#include "smtk/extension/matplotlib/Registrar.h"

#include "smtk/io/ReadMesh.h"

#include "smtk/model/DefaultSession.h"

#include "smtk/mesh/core/Resource.h"

#include "smtk/operation/Operation.h"

#include "smtk/plugin/Registry.h"

#include <fstream>
#include <streambuf>

//force to use filesystem version 3
#define BOOST_FILESYSTEM_VERSION 3
#include <boost/filesystem.hpp>

#include "smtkRegressionTestImage.h"
#include <vtkImageActor.h>
#include <vtkImageMapper3D.h>
#include <vtkPNGReader.h>
#include <vtkRenderWindow.h>
#include <vtkRenderer.h>
#include <vtkSmartPointer.h>
#include <vtkTesting.h>

namespace
{
std::string data_root = SMTK_DATA_DIR;
std::string write_root = SMTK_SCRATCH_DIR;

void cleanup(const std::string& file_path)
{
  //first verify the file exists
  ::boost::filesystem::path path(file_path);
  if (::boost::filesystem::is_regular_file(path))
  {
    //remove the file_path if it exists.
    ::boost::filesystem::remove(path);
  }
}
} // namespace

// Demonstrate/test the ability to call python SMTK operators within the C++
// environment.
int RenderMesh(int argc, char* argv[])
{
  (void)argc;
  (void)argv;

  // Create an operation manager
  smtk::operation::Manager::Ptr operationManager = smtk::operation::Manager::create();

  // Register matplotlib operators to the operation manager
  auto registry =
    smtk::plugin::addToManagers<smtk::extension::matplotlib::Registrar>(operationManager);

  // Access a 2-dimensional mesh with interesting z-features from the
  // data directory
  std::string input_path(data_root);
  input_path += "/mesh/2d/warpedMesh.h5m";

  smtk::mesh::ResourcePtr c = smtk::mesh::Resource::create();
  smtk::io::readMesh(input_path, c);

  if (!c || !c->isValid())
  {
    std::cerr << "Could not read mesh from " << input_path << std::endl;
    return 1;
  }

  // Construct a "render mesh" operator
  smtk::operation::Operation::Ptr op =
    operationManager->create("smtk.extension.matplotlib.render_mesh.RenderMesh");

  if (!op)
  {
    std::cerr << "Couldn't create \"render mesh\" operator" << std::endl;
    return 1;
  }

  // Set the mesh to be rendered
  op->parameters()->associate(smtk::mesh::Component::create(c->meshes()));

  // Set the file path for the rendered image
  std::string write_path(write_root);
  write_path += "/" + smtk::common::UUID::random().toString() + ".png";
  op->parameters()->findFile("filename")->setValue(write_path);

  // Execute the operator
  smtk::operation::Operation::Result result = op->operate();

  // Confirm that the operator succeeded
  if (
    result->findInt("outcome")->value() !=
    static_cast<int>(smtk::operation::Operation::Outcome::SUCCEEDED))
  {
    std::cerr << "render mesh operator failed\n";
    return 1;
  }

  // Read the image
  vtkSmartPointer<vtkPNGReader> reader = vtkSmartPointer<vtkPNGReader>::New();
  reader->SetFileName(write_path.c_str());

  // Visualize
  vtkSmartPointer<vtkImageActor> actor = vtkSmartPointer<vtkImageActor>::New();
  actor->GetMapper()->SetInputConnection(reader->GetOutputPort());

  vtkSmartPointer<vtkRenderer> renderer = vtkSmartPointer<vtkRenderer>::New();
  renderer->AddActor(actor);
  renderer->ResetCamera();

  vtkSmartPointer<vtkRenderWindow> renderWindow = vtkSmartPointer<vtkRenderWindow>::New();
  renderWindow->AddRenderer(renderer);

  // Compare the resulting image with the baseline
  if (
    smtkRegressionTestImage(renderWindow, 10., "matplotlib_rendered_mesh.png") !=
    vtkTesting::PASSED)
  {
    return 1;
  }

  // Remove the generated file
  cleanup(write_path);

  return 0;
}
